File: LR_2/lr2.py
#!/usr/bin/env python3
import os
import sys
import stat
from datetime import datetime
import argparse

# Размер буфера для копирования файлов (4 KB)
BUFFER_SIZE = 4096

def copy_file(source: str, dest: str) -> bool:
    # Копирование файла с использованием блочного чтения/записи.
    try:
        with open(source, 'rb') as src, open(dest, 'wb') as dst:
            while True:
                # Читаем блок данных
                buffer = src.read(BUFFER_SIZE)
                if not buffer:
                    break
                # Записываем блок в целевой файл
                dst.write(buffer)
        return True
    except IOError as e:
        print(f"Error copying file: {e}")
        return False

def move_file(source: str, dest: str) -> bool:
    # Перемещение файла.
    try:
        # Сначала пробуем просто переименовать
        os.rename(source, dest)
        return True
    except OSError:
        # Если не удалось, копируем и удаляем оригинал
        if copy_file(source, dest):
            try:
                os.remove(source)
                return True
            except OSError as e:
                print(f"Error removing source file: {e}")
                return False
        return False

def get_file_info(filename: str) -> None:
    #Получение и вывод информации о файле.
    try:
        file_stat = os.stat(filename)

        # Получаем права доступа в восьмеричном формате
        mode = stat.S_IMODE(file_stat.st_mode)

        # Форматируем время последней модификации
        mtime = datetime.fromtimestamp(file_stat.st_mtime)

        print(f"File: {filename}")
        print(f"Size: {file_stat.st_size} bytes")
        print(f"Permissions: {oct(mode)[2:]} ({stat.filemode(file_stat.st_mode)})")
        print(f"Last modified: {mtime}")

    except OSError as e:
        print(f"Error getting file information: {e}")

def change_permissions(filename: str, mode: str) -> bool:
    # Изменение прав доступа к файлу.
    try:
        # Преобразуем строку с восьмеричным числом в int
        mode_int = int(mode, 8)
        os.chmod(filename, mode_int)
        return True
    except (OSError, ValueError) as e:
        print(f"Error changing file permissions: {e}")
        return False

def console_interface():
    # Интерактивный консольный интерфейс программы.
    while True:
        print("\nFile Operations:")
        print("1. Copy file")
        print("2. Move file")
        print("3. Get file information")
        print("4. Change file permissions")
        print("5. Exit")

        try:
            choice = int(input("Choose operation (1-5): "))
        except ValueError:
            print("Please enter a number between 1 and 5")
            continue

        if choice == 5:
            break

        if choice == 1:
            source = input("Enter source file: ")
            dest = input("Enter destination file: ")
            if copy_file(source, dest):
                print("File copied successfully")

        elif choice == 2:
            source = input("Enter source file: ")
            dest = input("Enter destination file: ")
            if move_file(source, dest):
                print("File moved successfully")

        elif choice == 3:
            filename = input("Enter filename: ")
            get_file_info(filename)

        elif choice == 4:
            filename = input("Enter filename: ")
            mode = input("Enter new permissions (octal, e.g. 644): ")
            if change_permissions(filename, mode):
                print("Permissions changed successfully")

        else:
            print("Invalid choice")

def main():
    # Создаём парсер аргументов командной строки
    parser = argparse.ArgumentParser(description='File operations program')

    # Добавляем аргументы
    parser.add_argument('--copy', nargs=2, metavar=('SOURCE', 'DEST'),
                        help='Copy source file to destination')
    parser.add_argument('--move', nargs=2, metavar=('SOURCE', 'DEST'),
                        help='Move source file to destination')
    parser.add_argument('--info', metavar='FILE',
                        help='Display file information')
    parser.add_argument('--chmod', nargs=2, metavar=('FILE', 'MODE'),
                        help='Change file permissions (octal mode)')

    args = parser.parse_args()

    # Если нет аргументов, запускаем интерактивный режим
    if len(sys.argv) == 1:
        console_interface()
        return

    # Обработка аргументов командной строки
    if args.copy:
        if copy_file(args.copy[0], args.copy[1]):
            print("File copied successfully")

    elif args.move:
        if move_file(args.move[0], args.move[1]):
            print("File moved successfully")

    elif args.info:
        get_file_info(args.info)

    elif args.chmod:
        if change_permissions(args.chmod[0], args.chmod[1]):
            print("Permissions changed successfully")

if __name__ == "__main__":
    main()

File: LR_2/lr2.cpp
#include <iostream>
#include <fcntl.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string>
#include <cstring>
#include <time.h>

using namespace std;

void printHelp() {
    cout << "File Operations Program\n\n"
         << "Usage from command line:\n"
         << "  --copy source_file dest_file   : Copy file\n"
         << "  --move source_file dest_file   : Move file\n"
         << "  --info file_name               : Get file information\n"
         << "  --chmod file_name mode         : Change file permissions\n"
         << "  --help                         : Show this help\n\n"
         << "Usage from console:\n"
         << "  Run program without arguments and follow prompts\n";
}

// Размер буфера для блочного копирования файлов - 4KB
const int BUFFER_SIZE = 4096;

// Копирование файла с использованием блочного чтения/записи
bool copyFile(const char* source, const char* dest) {
    // Открываем исходный файл только для чтения
    int fd_source = open(source, O_RDONLY);
    if (fd_source == -1) {
        cout << "Error opening source file\n";
        return false;
    }

    // Создаем/перезаписываем целевой файл с правами 644 (rw-r--r--)
    int fd_dest = open(dest, O_WRONLY | O_CREAT | O_TRUNC, 0644);
    if (fd_dest == -1) {
        close(fd_source);
        cout << "Error creating destination file\n";
        return false;
    }

    char buffer[BUFFER_SIZE];
    ssize_t bytes_read;

    // Читаем блоками по BUFFER_SIZE байт, пока не достигнем конца файла
    while ((bytes_read = read(fd_source, buffer, BUFFER_SIZE)) > 0) {
        // Проверяем что записали столько же байт, сколько прочитали
        if (write(fd_dest, buffer, bytes_read) != bytes_read) {
            close(fd_source);
            close(fd_dest);
            cout << "Error writing to destination file\n";
            return false;
        }
    }

    close(fd_source);
    close(fd_dest);
    return bytes_read != -1;
}

// Получение информации о файле через системный вызов stat()
void getFileInfo(const char* filename) {
    struct stat file_stat;

    if (stat(filename, &file_stat) == -1) {
        cout << "Error getting file information\n";
        return;
    }

    // Выводим основные атрибуты файла
    cout << "File: " << filename << "\n"
         << "Size: " << file_stat.st_size << " bytes\n"
         << "Permissions: " << (file_stat.st_mode & 0777) << "\n"
         << "Last modified: " << ctime(&file_stat.st_mtime);
}

// Перемещение файла
bool moveFile(const char* source, const char* dest) {
    if (rename(source, dest) == -1) {
        if (copyFile(source, dest)) {
            return unlink(source) != -1;
        }
        return false;
    }
    return true;
}

// Изменение прав доступа к файлу
bool changePermissions(const char* filename, mode_t mode) {
    return chmod(filename, mode) != -1;
}

// Консольный интерфейс
void consoleInterface() {
    while (true) {
        cout << "\nFile Operations:\n"
             << "1. Copy file\n"
             << "2. Move file\n"
             << "3. Get file information\n"
             << "4. Change file permissions\n"
             << "5. Exit\n"
             << "Choose operation (1-5): ";

        int choice;
        cin >> choice;

        if (choice == 5) break;

        string source, dest;
        mode_t mode;

        switch (choice) {
            case 1:
                cout << "Enter source file: ";
                cin >> source;
                cout << "Enter destination file: ";
                cin >> dest;
                if (copyFile(source.c_str(), dest.c_str())) {
                    cout << "File copied successfully\n";
                }
                break;

            case 2:
                cout << "Enter source file: ";
                cin >> source;
                cout << "Enter destination file: ";
                cin >> dest;
                if (moveFile(source.c_str(), dest.c_str())) {
                    cout << "File moved successfully\n";
                }
                break;

            case 3:
                cout << "Enter filename: ";
                cin >> source;
                getFileInfo(source.c_str());
                break;

            case 4:
                cout << "Enter filename: ";
                cin >> source;
                cout << "Enter new permissions (octal): ";
                cin >> oct >> mode;
                if (changePermissions(source.c_str(), mode)) {
                    cout << "Permissions changed successfully\n";
                }
                break;

            default:
                cout << "Invalid choice\n";
        }
    }
}

int main(int argc, char* argv[]) {
    if (argc == 1) {
        consoleInterface();
        return 0;
    }

    string command = argv[1];

    if (command == "--help") {
        printHelp();
    }
    else if (command == "--copy" && argc == 4) {
        if (copyFile(argv[2], argv[3])) {
            cout << "File copied successfully\n";
        }
    }
    else if (command == "--move" && argc == 4) {
        if (moveFile(argv[2], argv[3])) {
            cout << "File moved successfully\n";
        }
    }
    else if (command == "--info" && argc == 3) {
        getFileInfo(argv[2]);
    }
    else if (command == "--chmod" && argc == 4) {
        mode_t mode;
        sscanf(argv[3], "%o", &mode);
        if (changePermissions(argv[2], mode)) {
            cout << "Permissions changed successfully\n";
        }
    }
    else {
        cout << "Invalid arguments. Use --help for usage information.\n";
    }

    return 0;
}